### ISSUE TITLE
Build/install HTCondor python bindings from PyPi

### ISSUE THREAD

**Issue by amaltaro** [dmwm/WMCore#10256]
**Impact of the new feature**
WMAgent (and any other software that depends on Condor)

**Is your feature request related to a problem? Please describe.**
It isn't really a WMCore issue, but it's easier to keep track of this task here than in the cms-sw/cmsdist repository.
Not a problem, but just trying to move away of the RPM deployment model.

**Describe the solution you'd like**
With the effort to mainstream our software deployment and migrate to python managers and containeirization of our software, we would benefit if we could build/deploy the HTCondor python bindings via `pip`, instead of building our own RPM based on the condor.spec file:
https://github.com/cms-sw/cmsdist/blob/comp_gcc630/condor.spec

The `htcondor` python bindings are already available in PyPi, and one can download the package wheels (binary). However, I tried to build it today for both python2 and python3, but it fails with some missing dependency that I had no courage to investigate further.

For the record, this is the initial spec file that I created (note the source type, since there are no source code available via Pypi. The Requires line might be dropped, needs to be investigated):
```
amaltaro@vocms055:/build/amaltaro $ cat cmsdist/py2-htcondor.spec 
### RPM external py2-htcondor 8.9.11
## IMPORT build-with-pip

Requires: zlib expat pcre libtool boost172 p5-archive-tar curl libxml2 p5-time-hires libuuid sqlite
%define PipDownloadSourceType none
```

**Describe alternatives you've considered**
Keep using the old condor RPM, which is much more complicated than using a pip install.

**Additional context**
The package is properly built, but installation fails with missing dependencies. This is the output in the terminal:
```
* The action "install-external+py2-htcondor+8.9.11" was not completed successfully because Traceback (most recent call last):
File "/build/amaltaro/pkgtools/scheduler.py", line 206, in doSerial
result = commandSpec[0](*commandSpec[1:])
File "pkgtools/cmsBuild", line 3436, in installPackage
File "pkgtools/cmsBuild", line 3188, in installRpm
RpmInstallFailed: Failed to install package py2-htcondor. Reason:
error: Failed dependencies:
	libcrypto-de8e2921.so.1.1(OPENSSL_1_1_0)(64bit) is needed by external+py2-htcondor+8.9.11-1-1.x86_64
	libcrypto-de8e2921.so.1.1(OPENSSL_1_1_0d)(64bit) is needed by external+py2-htcondor+8.9.11-1-1.x86_64
	libcrypto-de8e2921.so.1.1(OPENSSL_1_1_0f)(64bit) is needed by external+py2-htcondor+8.9.11-1-1.x86_64
	libcrypto-de8e2921.so.1.1(OPENSSL_1_1_0i)(64bit) is needed by external+py2-htcondor+8.9.11-1-1.x86_64
	libcrypto-de8e2921.so.1.1(OPENSSL_1_1_1)(64bit) is needed by external+py2-htcondor+8.9.11-1-1.x86_64
	libglobus_gssapi_gsi-46531449.so.4.10.9(globus_gssapi_gsi)(64bit) is needed by external+py2-htcondor+8.9.11-1-1.x86_64
	libgomp-3300acd3.so.1.0.0(GOMP_1.0)(64bit) is needed by external+py2-htcondor+8.9.11-1-1.x86_64
	libgomp-3300acd3.so.1.0.0(OMP_1.0)(64bit) is needed by external+py2-htcondor+8.9.11-1-1.x86_64
	libk5crypto-622ef25b.so.3.1(k5crypto_3_MIT)(64bit) is needed by external+py2-htcondor+8.9.11-1-1.x86_64
	libkeyutils-1-ff31573b.2.so(KEYUTILS_0.3)(64bit) is needed by external+py2-htcondor+8.9.11-1-1.x86_64
	libkrb5-fb0d2caa.so.3.3(krb5_3_MIT)(64bit) is needed by external+py2-htcondor+8.9.11-1-1.x86_64
	libkrb5support-d7ce89d4.so.0.1(krb5support_0_MIT)(64bit) is needed by external+py2-htcondor+8.9.11-1-1.x86_64
	libssl-6536aedd.so.1.1(OPENSSL_1_1_0)(64bit) is needed by external+py2-htcondor+8.9.11-1-1.x86_64
```

**Comment by amaltaro** [comment]
NOTE: if we do not build this package via Pypi, and instead keep using the current spec file where we build it all from scratch. We likely need to create a python3 version of that package/spec and make it depend on python3 setuptools, instead of the python2 dependency. I'm also not sure whether we could change that to `BuildRequires` only, and avoid getting python2 deployed with other things.

**Comment by amaltaro** [comment]
@smuzaffar Hi Shahzad, I resumed working on this item, thus trying to install a py3-htcondor.spec from Pypi. 
As expected, I hit again the same problem as the one reported in this GH issue:
```
RpmInstallFailed: Failed to install package py3-htcondor. Reason:
error: Failed dependencies:
	libcrypto-d47fc050.so.1.1(OPENSSL_1_1_0)(64bit) is needed by external+py3-htcondor+8.9.7-1-1.x86_64
	libcrypto-d47fc050.so.1.1(OPENSSL_1_1_0d)(64bit) is needed by external+py3-htcondor+8.9.7-1-1.x86_64
....
```

could you please guide me on how to fulfill these dependencies on pip-like specs?

**Comment by amaltaro** [comment]
Alternatively, I tried changing the current condor.spec file to actually depend on `python3` and `py3-setuptools`, instead of the python2 equivalent and updated these as well:
```
  -DPYTHON_EXECUTABLE:FILEPATH=${PYTHON_ROOT}/bin/python3.8 \
  -DPYTHON_INCLUDE_DIR:PATH=${PYTHON_ROOT}/include/python3.8 \
  -DPYTHON_LIBRARY:FILEPATH=${PYTHON_ROOT}/lib/libpython3.8.so \
```

but it fails during the build configure step. AFAICS, the failure comes from a mix of python2 and python3 libraries, since boost172.spec depends on `python` (py2).

**Comment by amaltaro** [comment]
Just a note, I can successfully build condor with these changes:
```
-Requires: zlib expat pcre libtool python boost172 p5-archive-tar curl libxml2 p5-time-hires libuuid sqlite py2-setuptools
-BuildRequires: cmake gcc
+Requires: zlib expat pcre libtool boost172 p5-archive-tar curl libxml2 p5-time-hires libuuid sqlite
+BuildRequires: cmake gcc python py2-setuptools
```

where python and setuptools are only required for building time.

**Comment by smuzaffar** [comment]
@amaltaro , are you picking up openssl from system or cms externals?

**Comment by amaltaro** [comment]
@smuzaffar Hi Shahzad, as of a 3 or 4 months ago, we stopped using our COMP OpenSSL in favor of the system one.

BTW, Kenyi just managed to build condor (+boost) depending on python3 instead of python2, but the final goal would still to fetch those from pip instead of building everything from scratch. But we seem to have patches and many tweaks in the spec, it isn't clear to me how easy that process would be.

**Comment by amaltaro** [comment]
Condor with python3 libraries has been made available via (thanks to Kenyi!): https://github.com/cms-sw/cmsdist/pull/6925

I suggest we keep this issue opened to make sure that we eventually migrate to Pypi condor python bindings.

**Comment by khurtado** [comment]
@amaltaro Should we change the priority of this issue to low and create (and close) another issue that reflects the python3 compatible build for condor in https://github.com/cms-sw/cmsdist/pull/6925 ?

**Comment by amaltaro** [comment]
@khurtado I think the medium priority is good, as I consider deployment and containeirization to become higher priority in Q3 or Q4.

Do you want to create a new GH issue for creating a python3 compatible condor spec? Such that we can properly associate and account for that? I'm fine either way ;)

**Comment by khurtado** [comment]
@amaltaro Sounds good.  I just created the issue:
https://github.com/dmwm/WMCore/issues/10497

**Comment by amaltaro** [comment]
Great, thanks Kenyi! Let me actually move this issue to Q3 for the moment.

**Comment by amaltaro** [comment]
A new `condorpy3.spec` spec file has been created to be used within our python3 software stack. So I'm moving this issue out of the "Python3 support" GH project.

**Comment by amaltaro** [comment]
Once we get to this - or maybe even before that - we should also upgrade condor python bindings to >= 9.1.2.
The SI team started updating the testbed nodes and it will soon be rolled to the production schedds as well.

With some luck, the new version would also fix this super annoying issue when draining agents: https://github.com/dmwm/WMCore/issues/10170

**Comment by amaltaro** [comment]
@khurtado Kenyi, as discussed in our team meeting today, I'm letting you deal with this issue ;)
Note that the target python bindings version might be newer by now (maybe 9.2.x)?

If possible, I would try to chase the PyPi approach here, if possible. However, if that takes more than a morning or so of work, we can likely keep our usual build setup and go "just" with a package version update. Please let me know what your thoughts are and/or if there are any questions to be clarified.

**Comment by khurtado** [comment]
@amaltaro It seems pip doesn't have source distributions for htcondor, only wheels.

So, this works:
```
$. pip3 download --no-cache-dir --no-deps  --disable-pip-version-check  -d . htcondor==9.2.0
Collecting htcondor==9.2.0
```

But this doesn't and you can see no non-binary versions are provided:
```
$pip3 download --no-cache-dir  --no-deps --no-binary=:all: --disable-pip-version-check  -d . htcondor==9.2.0
Collecting htcondor==9.2.0
  Could not find a version that satisfies the requirement htcondor==9.2.0 (from versions: )
No matching distribution found for htcondor==9.2.0
```

The second is what we do when using #build-with-pip3. Do you know why wheels are not used? It seems we won't be able to use pip for htcondor otherwise, since only wheels are provided via pip

**Comment by khurtado** [comment]
@smuzaffar Thanks! If I do that the `none` gets added as an input parameter, but the `--no-binary=:all` does disappear

```
Download command: cd /build/khurtado/condor3v2/w/tmp/tmpt2t81O; . /build/khurtado/condor3v2/w/slc7_amd64_gcc630/external/py3-pip/19.3.1/etc/profile.d/init.sh; pip3 download --no-cache-dir --no-deps  none --disable-pip-version-check  -d . htcondor==9.2.0; /bin/ls | grep -v files.list > files.list
```

**Comment by smuzaffar** [comment]
can you please share your spec?

**Comment by smuzaffar** [comment]
ok found it on vocms055. pleae remove 
```
#%define PipBuildOptions none
#``%define PipDownloadOptions none
```

even though these are in commented section but rpm expand these macros too and this causes  PipBuildOptions to be set none.  remove these and set `%define PipDownloadSourceType none`

**Comment by khurtado** [comment]
@smuzaffar Thank you! I will give it a try.

**Comment by khurtado** [comment]
@smuzaffar 

Ok, that did the trick and pip gets the package, but installRpm seems to fail because it doesn't find a bunch of dependencies. Adding openssl as dependency doesn't seem to help (I tried as Requires or BuildRequires). Do you know how to fix this issue?

On the other hand, I have a working version that doesn't depend on pip. I will try again tomorrow with the pip version or will make a PR with the working version I have otherwise.

```
Build successful condorpy3.
(<function buildPackage at 0x7efff8956398>, <Package name=condorpy3>, <scheduler.Scheduler object at 0x7efff37ce850>) done
Trying to install the rpm package external+condorpy3+9.2.0 just built.
*snip*
* The action "install-external+condorpy3+9.2.0" was not completed successfully because Traceback (most recent call last):
File "/build/khurtado/condor3v2/pkgtools/scheduler.py", line 223, in doSerial
result = commandSpec[0](*commandSpec[1:])
File "pkgtools/cmsBuild", line 3483, in installPackage
File "pkgtools/cmsBuild", line 3235, in installRpm
RpmInstallFailed: Failed to install package condorpy3. Reason:
error: Failed dependencies:
	libcrypto-b8807daa.so.1.1(OPENSSL_1_1_0)(64bit) is needed by external+condorpy3+9.2.0-1-1.x86_64
	libcrypto-b8807daa.so.1.1(OPENSSL_1_1_0d)(64bit) is needed by external+condorpy3+9.2.0-1-1.x86_64
	libcrypto-b8807daa.so.1.1(OPENSSL_1_1_0f)(64bit) is needed by external+condorpy3+9.2.0-1-1.x86_64
	libcrypto-b8807daa.so.1.1(OPENSSL_1_1_0i)(64bit) is needed by external+condorpy3+9.2.0-1-1.x86_64
	libcrypto-b8807daa.so.1.1(OPENSSL_1_1_1)(64bit) is needed by external+condorpy3+9.2.0-1-1.x86_64
	libglobus_gssapi_gsi-46531449.so.4.10.9(globus_gssapi_gsi)(64bit) is needed by external+condorpy3+9.2.0-1-1.x86_64
	libgomp-3300acd3.so.1.0.0(GOMP_1.0)(64bit) is needed by external+condorpy3+9.2.0-1-1.x86_64
	libgomp-3300acd3.so.1.0.0(OMP_1.0)(64bit) is needed by external+condorpy3+9.2.0-1-1.x86_64
	libk5crypto-622ef25b.so.3.1(k5crypto_3_MIT)(64bit) is needed by external+condorpy3+9.2.0-1-1.x86_64
	libkeyutils-1-ff31573b.2.so(KEYUTILS_0.3)(64bit) is needed by external+condorpy3+9.2.0-1-1.x86_64
	libkrb5-fb0d2caa.so.3.3(krb5_3_MIT)(64bit) is needed by external+condorpy3+9.2.0-1-1.x86_64
	libkrb5support-d7ce89d4.so.0.1(krb5support_0_MIT)(64bit) is needed by external+condorpy3+9.2.0-1-1.x86_64
	libssl-6536aedd.so.1.1(OPENSSL_1_1_0)(64bit) is needed by external+condorpy3+9.2.0-1-1.x86_64

* The action "final-job" was not completed successfully because The following dependencies could not complete:
install-external+condorpy3+9.2.0
```

**Comment by smuzaffar** [comment]
looks like the prebuild wheel is build using `openssl 1.1.X` version which system openssl version is `1.0.2`. So you can not use the wheel. you have  to download the sources and build locally

**Comment by amaltaro** [comment]
Fixed by https://github.com/cms-sw/cmsdist/pull/7368 , which provides condor `9.2.0`

As discussed above, we cannot use the condor python bindings from PyPi because those are pre-built with a different openssl version.
@khurtado is there anything else to add here? If not, shall we close it?

Thank you very much to you and Shahzad by the way!

**Comment by smuzaffar** [comment]
looks like all these missing libs/requires are part of htcondor pip package itself and `htcondor.so` uses `rpath` to load these. So @vkuznet is right, just add all these as `Provides: ...` in yourspec file. No need to set any extra `LD_LIBRARY_PATH`


[a]
```
> ls lib/python3.9/site-packages/htcondor.libs/
libclassad-f7e7cee1.so.9.2.0       libcrypto-b8807daa.so.1.1            libglobus_common-a907bfdd.so.0.18.0          libglobus_gsi_credential-93e70014.so.1.7.0  libglobus_gss_assist-65a6e3f6.so.3.9.1   libglobus_openssl-b30c37ee.so.0.5.0        libgomp-3300acd3.so.1.0.0       libkeyutils-1-ff31573b.2.so     libltdl-c1a91d42.so.7.3.1   libpyclassad3-ebaba290.9_9_2_0_wheel.so  libssl-6536aedd.so.1.1
libcom_err-beb60336.so.2.1         libexpat-72eb0598.so.0.5.0           libglobus_gsi_callback-9822f46e.so.0.6.0     libglobus_gsi_proxy_core-12a13129.so.0.9.1  libglobus_gssapi_gsi-46531449.so.4.10.9  libglobus_openssl_error-0b738a0c.so.0.4.0  libgssapi_krb5-174f8956.so.2.2  libkrb5-fb0d2caa.so.3.3         libmunge-ff99ca73.so.2.0.0  libselinux-cf8f9094.so.1                 libuuid-fbabf189.so.1.2
libcondor_utils_9_2_0-4aacfabe.so  libglobus_callout-f50ca9f8.so.0.4.0  libglobus_gsi_cert_utils-fdcaf639.so.0.10.1  libglobus_gsi_sysconfig-231cd042.so.1.8.1   libglobus_oldgaa-a96b6c72.so.0.6.0       libglobus_proxy_ssl-ddb90ee5.so.1.5.0      libk5crypto-622ef25b.so.3.1     libkrb5support-d7ce89d4.so.0.1  libpcre-84ecf085.so.1.2.11  libsepol-b4f5b513.so.1                   libvomsapi-e886b23f.so.1.0.0
```

[b]
```
> objdump -x lib/python3.9/site-packages/htcondor/htcondor.so  | grep RPATH
  RPATH                $ORIGIN/htcondor_libs:$ORIGIN/../htcondor.libs
```
